From 76e5c8ba21632ffc938558986688ad014dd87522 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Thu, 30 Sep 2021 16:16:11 +0300
Subject: [PATCH] [SW-57783] pci: disable iatu region before configuring

iATU regions might be reconfigured multiple times on runtime.
If a transaction is being translated by an iATU region while it is
being reconfigured, the destination address is undefined, and an
innocent transaction might become harmful to the system.
To avoid that undefined behaviour, we disable each iATU region prior
to changing its configuration.

Change-Id: If8c4971ce4b435cacf885725471ebcc4ff149038
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/lib/common/pci.c b/lib/common/pci.c
index 6f3e330..49290b7 100644
--- a/lib/common/pci.c
+++ b/lib/common/pci.c
@@ -102,6 +102,11 @@
 	/* region offset */
 	offset += (IATU_REGION_SIZE * region) + IATU_REGION_INBOUND_OFF;
 
+	/* disable region before configuring */
+	ctrl_reg_val = regmap_readl(offset + IATU_REGION_CTRL_2_OFF);
+	ctrl_reg_val &= ~IATU_REGION_EN_MASK;
+	regmap_writel(ctrl_reg_val, offset + IATU_REGION_CTRL_2_OFF);
+
 	if (pci_region->mode == PCI_ADDRESS_MATCH_MODE) {
 		hl_pci_get_bar_phys_addr(pci_region->bar, &bar_phys_base);
 		region_base = bar_phys_base + pci_region->offset_in_bar;
@@ -153,6 +158,9 @@
 
 	regmap_writel(IATU_ACCESS_ENABLE_VAL, mmPCIE_AUX_DBI);
 
+	/* disable region before configuring */
+	regmap_writel(0, offset + IATU_REGION_CTRL_2_OFF);
+
 	/* Outbound Region 0 */
 	outbound_region_end_address =
 			pci_region->addr + pci_region->size - 1;
